File: 108. Convert Sorted Array to Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return 
        elif len(nums) == 1:
            return TreeNode(nums[0])
        else:
            return self.helper(nums, 0, len(nums)-1)
    
    def helper(self, nums, lo, hi):
        if lo > hi or lo < 0 or hi > len(nums)-1:
            return 
        elif lo == hi:
            return TreeNode(nums[lo])
        else:
            mid = (hi - lo) // 2 + lo
            node = TreeNode(nums[mid])
            node.left = self.helper(nums, lo, mid-1)
            node.right = self.helper(nums, mid+1, hi)
        return node

File: 108. Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int[] nums;
    public TreeNode sortedArrayToBST(int[] nums) {
        this.nums = nums;
        return helper(0, nums.length-1);
    }
    private TreeNode helper(int st, int ed) {
        if (st > ed) {
            return null;
        } else if (st == ed) {
            return new TreeNode(nums[st]);
        } else {
            int mid = (ed-st)/2+st;
            TreeNode node = new TreeNode(nums[mid]);
            node.left = helper(st, mid-1);
            node.right = helper(mid+1, ed);
            return node;
        }
    }
}